packages/babel-plugin-transform-react-display-name/src/index.js

```diff
@@ -42,53 +69,161 @@ export default function ({ types: t }) {
     visitor: {
       ExportDefaultDeclaration({ node }, state) {
         if (isCreateClass(node.declaration)) {
-          let displayName = path.basename(state.file.opts.filename, path.extname(state.file.opts.filename));
+          const extname = pathMod.extname(state.file.opts.filename);
+          let displayName = pathMod.basename(state.file.opts.filename, extname);
 
           // ./{module name}/index.js
           if (displayName === "index") {
-            displayName = path.basename(path.dirname(state.file.opts.filename));
+            displayName = pathMod.basename(pathMod.dirname(state.file.opts.filename));
           }
 
-          addDisplayName(displayName, node.declaration);
+          addDisplayNameProperty(displayName, node.declaration);
         }
       },
-
       CallExpression(path) {
         const { node } = path;
         if (!isCreateClass(node)) return;
 
-        let id;
-
-        // crawl up the ancestry looking for possible candidates for displayName inference
-        path.find(function (path) {
-          if (path.isAssignmentExpression()) {
-            id = path.node.left;
-          } else if (path.isObjectProperty()) {
-            id = path.node.key;
-          } else if (path.isVariableDeclarator()) {
-            id = path.node.id;
-          } else if (path.isStatement()) {
-            // we've hit a statement, we should stop crawling up
-            return true;
-          }
-
-          // we've got an id! no need to continue
-          if (id) return true;
-        });
+        let id = findCandidateNameForExpression(path, true);
 
         // ensure that we have an identifier we can inherit from
         if (!id) return;
 
         // foo.bar -> bar
-        if (t.isMemberExpression(id)) {
+        if (types.isMemberExpression(id)) {
           id = id.property;
         }
 
         // identifiers are the only thing we can reliably get a name from
-        if (t.isIdentifier(id)) {
-          addDisplayName(id.name, node);
+        if (types.isIdentifier(id)) {
+          addDisplayNameProperty(id.name, node);
+        }
+      },
+      ClassDeclaration: function (path) {
+        if (classHasRenderMethod(path)) {
+          setDisplayNameAfter(path, path.node.id);
+        }
+      },
+      FunctionDeclaration: function (path, state) {
+        if (doesReturnJSX(path.node.body)) {
+          let displayName;
+          if (path.parentPath.node.type === "ExportDefaultDeclaration") {
+            if (path.node.id == null) {
+              // An anonymous function declaration in export default declaration.
+              // Transform `export default function () { ... }`
+              // to `var _uid1 = function () { .. }; export default __uid;`
+              // then add displayName to _uid1
+              const extension = pathMod.extname(state.file.opts.filename);
+              const name = pathMod.basename(state.file.opts.filename, extension);
```

Comment:
This could lead to an awful amount of components with `index` as the `displayName` 😄 

Though if you didn't use any heuristics at all they'd just be named the equally unhelpful `default`.